fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java

```diff
@@ -1169,5 +1178,33 @@ private void handleExportStmt() throws Exception {
     private List<PrimitiveType> exprToType(List<Expr> exprs) {
         return exprs.stream().map(e -> e.getType().getPrimitiveType()).collect(Collectors.toList());
     }
+
+    private void matchSql(QueryStmt queryStmt) throws AnalysisException {
+        Map<String, List<SqlBlockRule>> userToSqlBlockRuleMap = Catalog.getCurrentCatalog().getSqlBlocklistMgr().getUserToSqlBlockRuleMap();
+        // match default rule
+        String sql = queryStmt.getOrigStmt().originStmt;
+        List<SqlBlockRule> defaultRules = userToSqlBlockRuleMap.getOrDefault(SqlBlockRule.DEFAULT_USER, new ArrayList<>());
+        for (SqlBlockRule rule : defaultRules) {
+            matchSql(rule, sql);
+        }
+        // match user rule
+        String user = context.getUserIdentity().getUser();
+        List<SqlBlockRule> userRules = userToSqlBlockRuleMap.getOrDefault(user, new ArrayList<>());
+        for (SqlBlockRule rule : userRules) {
+            matchSql(rule, sql);
+        }
+    }
+
+    @VisibleForTesting
+    public static void matchSql(SqlBlockRule rule, String sql) throws AnalysisException {
+        if (rule.getEnable() != null && rule.getEnable()) {
+            String sqlPattern = rule.getSql();
+            Pattern pattern = Pattern.compile(sqlPattern);
```

Comment:
the pattern compilation should be done when the rule is created, so that we don't need to compile it every time.
You should be aware that this rule checking logic is on the critical path of query, so performance is top level consideration.

Reply:
Yes, there is a performance problem. Let me fix it